### DISCUSSION TITLE
Template matching on GPU & Dynamo

### DISCUSSION THREAD

**Discussion by JuhaHuiskonen** [KudryashevLab/TomoBEAR#14] [Ideas]
Is it possible to run tomoBEAR template matching (GPU), and write the results to Dynamo catalogue?  My project is now organised in a Dynamo catalogue, but the template matching (CPU) is too slow so I am looking for alternatives for this step. Thanks!

**Comment by MishaKudryashev** [comment]
Hi Juha, 
Yes, there are some ways to run this GPU-template matching, the simple one is a function in a for-loop. If you activate tomoBEAR and deactivate dynamo it would be something like this 

for i = 1:tomo_number
        gpuDevice(1);% allocate a device, numerates from 1 (GPU0 -> device 1 and so on)
        pts = dynamo_match(tomo,template,'mask',mask,'outputFolder',['TM_gpu/tomo_' num2strtotal(i,3)],'ytilt',maxangles(i,:),'sc',[1160 1200 1000],'cr',360,'cs',10, 'ir', 90, 'is', 10, 'bin',1);
        gpuDevice([]); %release GPU 
 end

It will produce you an output like dynamo_match on CPU - a variable that can be treated by dynamo particle extractor and the folders with the CC-volumes and angle-volumes.

Its a good idea to be able to import a dynamo Catalogue for GPU-based template matching and table generation, particle extraction, creation of the first project, etc. We will try to create a point of entry at this step. It will take some testing, we will write it here. 

Hope its useful,
MIsha

**Comment by JuhaHuiskonen** [comment]
Hi Misha,

Thanks, that seems pretty straightforward!  This is my current script for running dynamo_match.  It's not fully tested, but it seems to be doing what I want:  keep only views close to side views (60 <= tilt <=120), write a table, and then import the table to the current catalogue.

oc = dcm('c', 'warp_catalogue','l','t')
for i = 1:200
   fprintf('Starting template matching using tomogram %d', i);
   pts = dynamo_match(oc.results{i}, 'template.mrc', 'mask', 'mask.mrc', 'outputFolder', 'particles', 'ytilt', [-60,60], 'sc', [400,400,250], 'cr', 360, 'cs', 15, 'bin', 1, 'mw', 16);
   myTable=pts.peaks.computeTable('mcc',0.3);
   idx = myTable(:,8) >=60 & myTable(:,8) <= 120;
   newTable = myTable(idx,:);
   peaks_file = sprintf('tables/peaks_%d.tbl', i)
   model_name = sprintf('ccpeaks_%d', i)
   dwrite(newTable, peaks_file);
   if height(newTable) > 0
      m=dynamo_model_import('c', 'warp_catalogue', 'i', i, 'table', peaks_file, 'mn', model_name);
   end
end
 

Best wishes,
Juha

**Reply by MishaKudryashev** [reply]
Hi! 
Yes, I think you need to skip  'mw', 16 - its paralellzation in Dynamo. And you need to keep an eye that when you run dynamo_match you run it from tomoBear, for the other functions you may need to re-activate dynamo. 

Uh, its not pretty, we will think how to run it nicer. 

try it, hope it works!
Misha
